internal/uvm/create_lcow.go

```diff
@@ -0,0 +1,339 @@
+		if !opts.KernelDirect {
+			kernelArgs = "initrd=/" + opts.RootFSFile
+		}
+	case PreferredRootFSTypeVHD:
+		// Support for VPMem VHD(X) booting rather than initrd..
+		kernelArgs = "root=/dev/pmem0 ro init=/init"
+		imageFormat := "Vhd1"
+		if strings.ToLower(filepath.Ext(opts.RootFSFile)) == "vhdx" {
+			imageFormat = "Vhdx"
+		}
+		doc.VirtualMachine.Devices.VirtualPMem.Devices = map[string]hcsschema.VirtualPMemDevice{
+			"0": {
+				HostPath:    filepath.Join(opts.BootFilesPath, opts.RootFSFile),
+				ReadOnly:    true,
+				ImageFormat: imageFormat,
+			},
+		}
+		if err := wclayer.GrantVmAccess(uvm.id, filepath.Join(opts.BootFilesPath, opts.RootFSFile)); err != nil {
+			return nil, fmt.Errorf("failed to grantvmaccess to %s: %s", filepath.Join(opts.BootFilesPath, opts.RootFSFile), err)
+		}
+		// Add to our internal structure
+		uvm.vpmemDevices[0] = vpmemInfo{
+			hostPath: opts.RootFSFile,
+			uvmPath:  "/",
+			refCount: 1,
+		}
+	}
+
+	vmDebugging := false
+	if opts.ConsolePipe != "" {
+		vmDebugging = true
+		kernelArgs += " 8250_core.nr_uarts=1 8250_core.skip_txen_test=1 console=ttyS0,115200"
+		doc.VirtualMachine.Devices.ComPorts = map[string]hcsschema.ComPort{
+			"0": { // Which is actually COM1
+				NamedPipe: opts.ConsolePipe,
+			},
+		}
+	} else {
+		kernelArgs += " 8250_core.nr_uarts=0"
+	}
+
+	if opts.EnableGraphicsConsole {
+		vmDebugging = true
+		kernelArgs += " console=tty"
+		doc.VirtualMachine.Devices.Keyboard = &hcsschema.Keyboard{}
+		doc.VirtualMachine.Devices.EnhancedModeVideo = &hcsschema.EnhancedModeVideo{}
+		doc.VirtualMachine.Devices.VideoMonitor = &hcsschema.VideoMonitor{}
+	}
+
+	if !vmDebugging {
+		// Terminate the VM if there is a kernel panic.
+		kernelArgs += " panic=-1 quiet"
+	}
+
+	if opts.KernelBootOptions != "" {
+		kernelArgs += " " + opts.KernelBootOptions
+	}
+
+	// Start GCS with stderr pointing to the vsock port created below in
+	// order to forward guest logs to logrus.
+	initArgs := fmt.Sprintf("/bin/vsockexec -e %d /bin/gcs -log-format json -loglevel %s",
+		linuxLogVsockPort,
+		logrus.StandardLogger().Level.String())
+
+	if vmDebugging {
+		// Launch a shell on the console.
+		initArgs = `sh -c "` + initArgs + ` & exec sh"`
+	}
+
+	kernelArgs += ` pci=off brd.rd_nr=0 pmtmr=0 -- ` + initArgs
+
+	if !opts.KernelDirect {
+		doc.VirtualMachine.Chipset.Uefi = &hcsschema.Uefi{
+			BootThis: &hcsschema.UefiBootEntry{
+				DevicePath:   `\` + opts.KernelFile,
+				DeviceType:   "VmbFs",
+				OptionalData: kernelArgs,
+			},
+		}
+	} else {
+		doc.VirtualMachine.Chipset.LinuxKernelDirect = &hcsschema.LinuxKernelDirect{
+			KernelFilePath: filepath.Join(opts.BootFilesPath, opts.KernelFile),
+			KernelCmdLine:  kernelArgs,
+		}
+		if *opts.PreferredRootFSType == PreferredRootFSTypeInitRd {
+			doc.VirtualMachine.Chipset.LinuxKernelDirect.InitRdPath = filepath.Join(opts.BootFilesPath, opts.RootFSFile)
+		}
+	}
+
+	fullDoc, err := mergemaps.MergeJSON(doc, ([]byte)(opts.AdditionHCSDocumentJSON))
```

Comment:
May be a minor optimisation here by only calling merge if there is additional JSON. Saves a lot of work unmarshalling and marshalling. (Not tested). Same applies to WCOW, but strictly relevant in this PR.

Reply:
I wonder if we should get rid of this. No one is calling it...

Reply:
Docker still uses it for debug purposes right?

Reply:
But not on this codepath. Docker is v1. Question I guess comes down to whether you envisage this as another annotation just in case we haven't thought of something.

Reply:
Ah your right. Docker is on v1. I think we should remove it. Annotations (UVM opts) are a much better match to our code. Makes it easier to just add one-offs rather than needing to know the JSON structures

Reply:
Leaving it for now. And actually `MergeJSON` just returns if `len(additionalJSON) == 0` so we already skip this step today.